merge/multiple_appliers_test.go

```diff
@@ -243,6 +244,300 @@ func TestMultipleAppliersSet(t *testing.T) {
 	}
 }
 
+var structMultiversionParser = func() Parser {
+	parser, err := typed.NewParser(`types:
+- name: v1
+  map:
+    fields:
+      - name: struct
+        type:
+          namedType: struct
+      - name: version
+        type:
+          scalar: string
+- name: struct
+  map:
+    fields:
+      - name: name
+        type:
+          scalar: string
+      - name: scalarField_v1
+        type:
+          scalar: string
+      - name: complexField_v1
+        type:
+          namedType: complex
+- name: complex
+  map:
+    fields:
+      - name: name
+        type:
+          scalar: string
+- name: v2
+  map:
+    fields:
+      - name: struct
+        type:
+          namedType: struct_v2
+      - name: version
+        type:
+          scalar: string
+- name: struct_v2
+  map:
+    fields:
+      - name: name
+        type:
+          scalar: string
+      - name: scalarField_v2
+        type:
+          scalar: string
+      - name: complexField_v2
+        type:
+          namedType: complex_v2
+- name: complex_v2
+  map:
+    fields:
+      - name: name
+        type:
+          scalar: string
+- name: v3
+  map:
+    fields:
+      - name: struct
+        type:
+          namedType: struct_v3
+      - name: version
+        type:
+          scalar: string
+- name: struct_v3
+  map:
+    fields:
+      - name: name
+        type:
+          scalar: string
+      - name: scalarField_v3
+        type:
+          scalar: string
+      - name: complexField_v3
+        type:
+          namedType: complex_v3
+- name: complex_v3
+  map:
+    fields:
+      - name: name
+        type:
+          scalar: string
+`)
+	if err != nil {
+		panic(err)
+	}
+	return parser
+}()
+
+func TestMultipleAppliersFieldUnsetting(t *testing.T) {
+	versions := []fieldpath.APIVersion{"v1", "v2", "v3"}
+	for _, v1 := range versions {
+		for _, v2 := range versions {
+			for _, v3 := range versions {
+				t.Run(fmt.Sprintf("%s-%s-%s", v1, v2, v3), func(t *testing.T) {
+					testMultipleAppliersFieldUnsetting(t, v1, v2, v3)
+				})
+			}
+		}
+	}
+}
+
+func testMultipleAppliersFieldUnsetting(t *testing.T, v1, v2, v3 fieldpath.APIVersion) {
+	tests := map[string]TestCase{
+		"unset_scalar_sole_owner": {
+			Ops: []Operation{
+				Apply{
```

Comment:
Can we add a test that shares ownership with an updater too?

Reply:
Added the test. It failed so I poked around and found that the merge logic found:

https://github.com/kubernetes-sigs/structured-merge-diff/blob/068223b26491dad6419fc6712b70a57e0a162ef1/merge/update.go#L249

This seems to suggest that the merge logic is designed to drop associative map/list items from updaters rather than merge it in as it does with other appliers. Now that I'm also having fields be merged using the same logic, this is more problematic, because it would not allow a field like "replica" to be ownership transferred from an applier to an updater.

And thinking about it, it also seems like the wrong thing to do for associative map/list items since an applier might with to ownership transfer those to an updater as well... So I've updated the `appliers_remove_from_controller_real_conversion` test to reflect the behavior change.

But I'm not 100% comfortable with this because I don't understand the reasoning behind why the merge logic would be that way in the first place.

Reply:
Figured this out.

Offline @apelisse showed me one of the cases we want to preserve the behavior of:

- Applier creates a Deployment with a Container in the PodSpec
- Some controller modifies something about the container using update (not apply)
- The Applier removes the container from their applied configuration
- Desired result: the container is deleted

This case works fine with this PR. It does exactly what we want.

The reason the `appliers_remove_from_controller_real_conversion` test fails the object is first created with an updater, which then owns all the maps, from the root down, which doesn't reflect what we want to test for.  If I change the test so that the applier creates the object initially, and creates map a and b, then the test passes.

One of my misunderstandings here is how updater ownership works. They don't own all fields of an object, only fields they have changed.

Reply:
I've added an `applier_updater_shared_ownership_real_conversion` test and updated the `appliers_remove_from_controller_real_conversion` to show that this PR does what we want. This ensures both cases:

- An applier creates an object, a controller changes something about an associative map item, the applier removes the item from its applied configuration and the item is deleted
- An updater creates an object, an applier changes something about an associative map item, the applier removes the item from its applied configuration and the item is not deleted (without this PR, the item is deleted, which is not what I would expect or want)